pkg/reftracker/ref_tracker.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package reftracker
+
+import (
+	"fmt"
+	"sync"
+)
+
+type AppRefTracker struct {
+	lock       sync.Mutex
+	refsToApps map[RefKey]map[RefKey]struct{}
+	appsToRefs map[RefKey]map[RefKey]struct{}
+}
+
+func NewAppRefTracker() *AppRefTracker {
+	return &AppRefTracker{refsToApps: map[RefKey]map[RefKey]struct{}{}, appsToRefs: map[RefKey]map[RefKey]struct{}{}}
+}
+
+func (a AppRefTracker) AppsForRef(refKey RefKey) (map[RefKey]struct{}, error) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	apps := a.refsToApps[refKey]
+	if apps == nil {
+		return nil, fmt.Errorf("could not find ref %s", refKey.Description())
+	}
+
+	return apps, nil
+}
+
+func (a AppRefTracker) RefsForApp(appKey RefKey) (map[RefKey]struct{}, error) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	if a.appsToRefs[appKey] == nil {
+		return nil, fmt.Errorf("could not find refs for App %s", appKey.RefName())
+	}
+
+	return a.appsToRefs[appKey], nil
+}
+
+func (a AppRefTracker) RemoveRef(refKey RefKey) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	delete(a.refsToApps, refKey)
+}
+
+func (a AppRefTracker) RemoveAppFromAllRefs(appKey RefKey) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	refKeys := a.appsToRefs[appKey]
+	for refKey := range refKeys {
+		apps := a.refsToApps[refKey]
+		if apps == nil {
+			continue
+		}
+
+		delete(apps, appKey)
+		a.refsToApps[refKey] = apps
+	}
+
+	delete(a.appsToRefs, appKey)
+}
+
+func (a AppRefTracker) ReconcileRefs(currentRefs map[RefKey]struct{}, appKey RefKey) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// Add all new refs to AppRefTracker
+	for refKey := range currentRefs {
+		apps := a.refsToApps[refKey]
+		if apps == nil {
+			apps = map[RefKey]struct{}{}
+		}
+
+		appKey := NewAppKey(appKey.RefName(), refKey.Namespace())
```

Comment:
i dont think i quite understand why we need to "rebuild" app key. might have to zoom about it.